File: backend/app/schemas/subject.py
from pydantic import BaseModel


class PrerequisiteOut(BaseModel):
    id: int
    prerequisite_subject_id: int

    model_config = {"from_attributes": True}


class SubjectCreate(BaseModel):
    nome: str
    codigo: str
    ementa: str | None = None
    bibliografia: str | None = None
    resumo: str | None = None
    periodo_recomendado: int | None = None

    course_ids: list[int]
    prerequisite_ids: list[int] = []

class CourseRef(BaseModel):
    id: int
    nome: str

    model_config = {"from_attributes": True}

class SubjectOut(BaseModel):
    id: int
    nome: str
    codigo: str
    ementa: str | None = None
    bibliografia: str | None = None
    resumo: str | None = None
    periodo_recomendado: int | None = None

    prerequisitos: list[PrerequisiteOut] = []

    courses: list[CourseRef] = []

    model_config = {"from_attributes": True}

File: frontend/src/services/subject.ts
import { api } from "./api";

export interface Prerequisite {
  id: number;
  prerequisite_subject_id: number;
}

export interface Subject {
  id: number;
  course_id: number;
  nome: string;
  codigo: string;
  ementa: string | null;
  bibliografia: string | null;
  resumo: string | null;
  periodo_recomendado: number | null;
  prerequisitos: Prerequisite[];
}

export const subjectService = {
  list: (params?: { course_id?: number; search?: string }) => {
    const query = new URLSearchParams();
    if (params?.course_id) query.set("course_id", String(params.course_id));
    if (params?.search) query.set("search", params.search);
    const qs = query.toString();
    return api.get<Subject[]>(`/subjects/${qs ? `?${qs}` : ""}`);
  },

  getById: (id: number) => api.get<Subject>(`/subjects/${id}`),

  create: (data: {
    course_id: number;
    nome: string;
    codigo: string;
    ementa?: string;
    bibliografia?: string;
    resumo?: string;
    periodo_recomendado?: number;
  }) => api.post<Subject>("/subjects/", data),
};
